### PR TITLE
fix(sync): quieter logs when offline

### PR DESCRIPTION
**Pull request by mattslight**
## The noise

Came back to a laptop after wifi was off overnight and found ~30 lines × N retries of:

\`\`\`
[memory] pull failed: TypeError: fetch failed
    at node:internal/deps/undici/undici:13510:13
    ...
    [cause]: Error: getaddrinfo ENOTFOUND cloud.oyster.to
\`\`\`

Behaviour was correct (caught, no crash, retry every ~30 s) but the terminal was overwhelmed.

## The fix

New \`server/src/sync-log.ts\`:

- **\`formatSyncError(err)\`** — recognises ENOTFOUND, EAI_AGAIN, ECONNREFUSED, ECONNRESET, ENETUNREACH, ETIMEDOUT, UND_ERR_CONNECT_TIMEOUT, UND_ERR_SOCKET and returns \`cloud unreachable (CODE)\`. Non-network errors fall through to the original message.
- **\`createOfflineLogger(prefix)\`** — dedupe wrapper:
  - First failure → one concise line + "further offline retries suppressed"
  - Subsequent identical failures → silent
  - Heartbeat every 30 (default) → \`still offline (N attempts)\`
  - First success after streak → \`back online (after N failed attempts)\`
  - Non-network errors → always logged in full, counter untouched (preserves debug)

Wired into \`memory-sync-service\` (push + pull) and \`session-sync-service\` (push + pull + pushBytes). One logger per call-site.

## Before / after

**Before** (30+ minutes offline = N × 30 lines):
\`\`\`
[memory] pull failed: TypeError: fetch failed at node:internal/deps/undici/...
    ... 28 more lines ...
[sessions] pull failed: TypeError: fetch failed at node:internal/deps/undici/...
    ... 28 more lines ...
[memory] pull failed: TypeError: fetch failed at node:internal/deps/undici/...
    ...
\`\`\`

**After** (same period):
\`\`\`
[memory] pull cloud unreachable (ENOTFOUND) (further offline retries suppressed)
[sessions] pull cloud unreachable (ENOTFOUND) (further offline retries suppressed)
[memory] pull still offline (30 attempts) — cloud unreachable (ENOTFOUND)
[sessions] pull still offline (30 attempts) — cloud unreachable (ENOTFOUND)
... eventually ...
[memory] pull back online (after 47 failed attempts)
[sessions] pull back online (after 47 failed attempts)
\`\`\`

## Test plan

- [x] 14 new unit tests for \`formatSyncError\` / \`isOfflineLikeError\` / \`createOfflineLogger\` covering classification, dedupe, heartbeat cadence, recovery, singular/plural wording, and the "non-offline error preserves the counter" edge case
- [x] 260 server tests pass (was 246, +14)
- [x] tsc + \`npm run build\` clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: server/src/session-sync-service.ts [hunk 1]
```diff
@@ -477,7 +485,7 @@ export function createSessionSyncService(deps: SessionSyncDeps): SessionSyncServ
             { method: "PUT", headers, body: chunkBody },
           );
         } catch (err) {
-          console.warn(`[sessions] pushBytes chunk ${nextChunkNumber} fetch failed for ${sessionId}:`, err);
+          pushBytesLog.failure(err);
           break;
         }
```
[the remark this answers is not in the document]
Comment (mattslight) [reply server/src/session-sync-service.ts:490]: Fixed in 3c83baf — extended `OfflineLogger.failure(err, context?)` with an optional context string. The first-occurrence line and any non-offline error line include it; heartbeats stay generic (by then the streak may belong to multiple resources). Wired `session=c90da198 chunk=2` at the pushBytes call site.

File: server/src/sync-log.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  "UND_ERR_CONNECT_TIMEOUT", // undici-specific: TCP connect timeout
+  "UND_ERR_SOCKET",       // undici-specific: socket error
+]);
+
+/** Format an error for logging. For known transient network errors,
+ *  return a concise one-liner naming the code. For everything else, fall
+ *  through to the full message (callers can still console.warn with the
+ *  raw err object for a stack trace on real bugs). */
+export function formatSyncError(err: unknown): string {
+  if (!err) return "unknown error";
```
[the remark this answers is not in the document]
Comment (mattslight) [reply server/src/sync-log.ts]: Fixed in 3c83baf — `err == null` check now. `0`, `false`, `""` fall through to `String(err)`. Added test covering null/undefined/0/false/"".

File: server/src/sync-log.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+/** Create a dedupe-aware logger for a single sync operation. Holds a
+ *  consecutive-failures counter in closure; one logger per call-site so
+ *  pull and push (and bytes-push and reassemble) all track independently. */
+export function createOfflineLogger(
+  prefix: string,
+  options: OfflineLoggerOptions = {},
+): OfflineLogger {
+  const heartbeatEvery = options.heartbeatEvery ?? 30;
```
[the remark this answers is not in the document]
Comment (mattslight) [reply server/src/sync-log.ts]: Fixed in 3c83baf — clamped to positive integer via `Number.isFinite(x) && x >= 1`; 0/-5/NaN/Infinity all fall back to the default. Test loops through each bad value and asserts the default heartbeat (30) still fires.

File: server/test/sync-log.test.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  it("false on bare Error with no code", () => {
+    expect(isOfflineLikeError(new Error("nope"))).toBe(false);
+  });
+});
+
+describe("createOfflineLogger", () => {
+  const warnSpy = vi.spyOn(console, "warn").mockImplementation(() => {});
+  afterEach(() => {
+    warnSpy.mockClear();
+  });
```
[the remark this answers is not in the document]
Comment (mattslight) [reply server/test/sync-log.test.ts:77]: Fixed in 3c83baf — added `afterAll(() => warnSpy.mockRestore())` so the spy doesn't leak past this test file.
